### ISSUE TITLE
How to get map from xiaomi servers?

### ISSUE THREAD

**Issue by messimm** [marcelrv/XiaomiRobotVacuumProtocol#42]
Is it possible to get mapping from Xiaomi server?
I know my region: sg. 
I send GET-request to https://sg.api.io.mi.com/app/home/getmapfileurl with obj_name parameter which is the same with result of get_map_v1 method. It doesn't work. I try to concatenate roboroommap%2FXXXXXXXX%2FX (X from digits) and URL. It doesn't work too. What Do I do in wrong way?

**Comment by marcelrv** [comment]
yes, it is possible, but you need to make sure the request is properly signed with the security keys
In the response you get the right (temporary)  URL where the robotmap can be downloaded. It is to simplistic to just try to try to download the roboroommap%2FXXXXXXXX%2FX  file directly from the server, without the required cookies/signatures.
You can look into openHAB source to get idea of the request to make.

**Comment by messimm** [comment]
Thank you for reply! Sorry, one more question. How do obstacles recognition work? Do it depend on vacuum cleaner model? Or it is result of processing by Xiaomi servers?

**Comment by dsbabkov** [comment]
Help me please. How should I concatenate urls?

I have roborock S6 and got in get_map_v1 response rubysc1%2FXXXXXXXXX%2FX

My region server in MI Home is Russian. Vacuum was removed and added again to MI Home after setting the region. Seems my url is https://ru.api.io.mi.com/app/home/getmapfileurl

I have no experience of requests... I have tryed to input to adress line of chrome https://ru.api.io.mi.com/app/home/getmapfileurl/rubysc1/XXXXXXXXX/X and https://ru.api.io.mi.com/app/home/getmapfileurl/rubysc1%2FXXXXXXXXX%2FX but server responses "404 not found" ...

**Comment by gjuhasz86** [comment]
I have been experimenting with this recently, and [created a gist](https://gist.github.com/gjuhasz86/b05333830cc616b70869a324350691ad) that can help downloading the map.
The basic underlying mechanism is that you first send the requests to log into the Xiaomi account and obtain the service token as well as other security keys and cookies, then make a POST request to getmapfileurl with the map name and some other parameters encoded in the url, then that service will reply with the actual url that you can use to get the map data.
If you start with the gist mentioned above, you can find out all the details by following the calls (it uses a python api).
